library/std/src/thread/scoped.rs

```diff
@@ -163,10 +165,10 @@ impl<'env> Scope<'env> {
     /// to recover from such errors.
     ///
     /// [`join`]: ScopedJoinHandle::join
-    pub fn spawn<'scope, F, T>(&'scope self, f: F) -> ScopedJoinHandle<'scope, T>
+    pub fn spawn<F, T>(&'scope self, f: F) -> ScopedJoinHandle<'scope, T>
     where
-        F: FnOnce(&Scope<'env>) -> T + Send + 'env,
-        T: Send + 'env,
+        F: FnOnce() -> T + Send + 'scope,
+        T: Send + 'scope,
```

Comment:
This shouldn't need to be as small as `'scope`
```suggestion
        T: Send + 'env,
```
mainly, I'd find it less problematic if `s.spawn(|| s)` were denied.

Reply:
`s.spawn(|| s)` seems to be denied in both cases.

Do you have an example of any difference that this change makes?

Reply:
Hmm, that was unexpected; it looks like the edition-2018 magic fails when the capture is used "by value" but with fully inferred types (the `*&& -> &` place flattening does not seem to occur if the `*&&` place starts off inferred). This includes a move in return position.

Thus, using `move || s` or `|| -> &_ { s }` are both examples that do pass when `T` is only restricted to be `: 'scope`.

  - [Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=40b548412eb66079b07c009027ef9be7)

Reply:
Is it a problem if `s.spawn(move || s)` works? I can't really think of a good use case, but I'd prefer to not unnecessarily add more restrictions to the return type than necessary. I also think the signature is easier to follow if every lifetime is `'scope` without having to wonder why `F` is `'scope` but `T` is `'env`.

Reply:
🤔 It all boils down to what does the "scoped runtime" do with the returned value when not joined / and potentially when leaked —I suspect the latter is the only actual area of problems, and the current implementation leaks stuff in that case, so at first glance it looks like it's gonna be fine.

  - If `s` can be returned, then `scopeguard::guard(s, |s| s.spawn(…))` could also be the value of type `T`, and then it _could_ be problematic to have such drop logic run within the runtime's auto-join cleanup _etc._. Basically by using `'scope` we are giving more flexibility to the caller, who oughtn't realistically need it, and which could impair a bit future changes to the implementation so as not to cause unsoundness within that contrived use case.

  - So, _conservatively_, using `'env` is the more cautious and future-proof approach; we could always loosen it afterwards. But I agree I haven't delved too much into "all potential cleanup implementations", and that it is thus also possible that there be no problem whatsoever with that extra flexibility.

To recap:

  - there _could be_ future-proofing, soundness-wise, issues with the more lenient `: 'scope` bound (but it may also be fine);
  - the signature _is_ a bit more complex with `F : 'scope` on one end, and `T : 'env` on the other.

So, one thing that can be done is sticking to `T : 'scope` for now, but adding a mention to this to the tracking issue so that we remember to think about it before stabilization 🙂

Reply:
I spent some more time thinking about this, and concluded that using `'scope` is actually *unsound* with the current implementation, since the active thread counter is decremented before dropping an ignored return value:

```rust
#![feature(scoped_threads)]

use std::{
    thread::{scope, sleep},
    time::Duration,
};

fn main() {
    let mut a = "asdf".to_string();
    let r = &a;
    scope(|s| {
        s.spawn(move || {
            scopeguard::guard(s, move |s| {
                sleep(Duration::from_millis(100));
                s.spawn(move || dbg!(r));
            })
        });
    });
    drop(a);
    sleep(Duration::from_millis(200));
}
```
With std from this PR:
```
[src/main.rs:15] r = "�K\u{f}\\"
```

Eep!

However! After thinking a bit more, I concluded that using `'env` doesn't fix things. It is *still* unsound with `'env`, and in fact even the code before the change in this PR is unsound:

```rust
fn main() {
    let mut a = "asdf".to_string();
    let r = &a;
    scope(|s| {
        s.spawn(move || {
            scopeguard::guard(r, move |r| {
                sleep(Duration::from_millis(100));
                dbg!(r);
            })
        });
    });
    drop(a);
    sleep(Duration::from_millis(200));
}
```
On Rust rustc 1.61.0-nightly (8769f4ef2 2022-03-02):
```
[src/main.rs:15] r = "!\u{1a}\u{7f}c"
```

This makes me think that it's unlikely that an implementation would be unsound with the `'scope` bound, but not with the `'env` bound, since it need to handle this last example anyway.

(The fix is to drop the return value before marking a thread as joined. So, an implementation issue, not an API issue. I'll send a PR for that.)

Reply:
Good catch! :slightly_smiling_face:

___

So, I've been thinking about this, to fully spell out what my hunch / feeling of uneasyness was.

Let's start with the following new names, for the remainder of this post:

 - `: 'env` is `: CanReferToSuffOutsideTheScopeButNotTheSpawner`:
     - renamed `'cannot_capture_inside_scope` in what follows
 - `'scope` is `CanAlsoSpawnScopedThreads`
     - renamed `'cannot_capture_inside_scope_except_spawner` in what follows (together with `Scope` becoming `ScopedSpawner`).

So from a start point of `F` and `T` both being `CanReferToSuffOutsideTheScopeButNotTheSpawner`, we loosen it down to `CanAlsoSpawnScopedThreads` capability to `F`, since we are removing the raw `spawner` handle it used to receive as a parameter.

```diff
- F : 'cannot_capture_inside_scope + FnOnce(ScopedSpawner...) -> T,
+ F : 'cannot_capture_inside_scope_except_spawner + FnOnce() -> T, // equivalent API
```

  - (This change is actually a very good summary of this whole PR, btw :smile:)

Now, you are also giving the `CanAlsoSpawnScopedThreads` capability to the return value of the closure.

```diff
- T : 'cannot_capture_inside_scope,
+ T : 'cannot_capture_inside_scope_except_spawner, // new capability
```

  - If such value is obtained through an explicit `.join()`, then that's pointless since the `spawner` handle itself is also available.

  - Else, the return value is no longer accessible to the user, and thus its sole API being called, if any, will be its `drop` glue.
      - If the value is leaked, then such a capability was pointless as well, since not called.
      - Else, the value is dropped around the join-point (but before it, once your PR fixing the soundness bug is merged).

So what granting it the `CanAlsoSpawnScopedThreads` capability achieves is allowing this implicit drop, right before the join point cleanup, to spawn extra scoped threads, leading to thread count changes etc.

  - This is an extra feature of the API, so it can be beneficial if we deem it to be a useful feature to have;
  - It's also a maintenance burden on the scoped threads API, since any time the join logic is touched, implementors will need to be mindful of that rough edge:

    ```rs
    impl<'spawner, T> Drop for Packet<'spawner, T> {
        fn drop(&mut self) {
            // ...
            {
                // Drop the closure's return value: THIS MAY SPAWN EXTRA THREADS
                *self.result.get_mut() = None;
            // ...
            self.spawner.decrement_num_running_threads(...)
    ```

With the caveat that mistakes here endanger soundness, and that removing this `CanAlsoSpawnScopedThreads` capability would be a breaking change.

___

  - So, it's not, granted, _that much_ of extra cognitive burden when dealing with the implementation, but it is _some_.
  - However, the obtained flexibility does feel incredibly niche, imho (moreover, even if I were wrong and that capability were useful, then we could always add it later on: it wouldn't be a breaking change).
  - Remains the argument of "symmetry in the lifetime bounds, for a simpler function signature", which to me is outweighed by the  "caution, extra threads may be spawned during drop glue" subtlety for implementors.

Reply:
> ```
> // Drop the closure's return value: THIS MAY SPAWN EXTRA THREADS
> ```

I wonder if there's any reasonable implementation where this requires special attention. The point I was trying to make towards the end of my last comment is that implementations need to take care of handling Drop implementations that borrow from `'env` anyway, which means the drop needs to happen before any kind of 'joining' happens, which also covers the case where someone spawns more threads from the Drop impl.

I don't think there is really any difference between spawning threads near the end of `F`, versus doing things in `T`s `Drop` impl, from the perspective of what needs to be taken into account for soundness/correctness. In both cases it's just a thread spawning more threads before it gets joined.

Reply:
> The point I was trying to make

Yeah I guess I didn't get it 100% the first time 😅; I remained convinced it would lead to a more subtle approach, but as you've patiently mentioned, the subtlety is already there even with `'env` 👍 Thanks for having put up with my initial skepticism! 🙏